### DISCUSSION TITLE
dehydrateStateUsing with repeater results in both modified and unmodified values

### DISCUSSION THREAD

**Discussion by RichardHuntCCI** [filamentphp/filament#6907] [Help] [answered]
### Context

We are creating dynamic forms to represent Terraform data structures. The forms schemas are generated based on the structure strings, e.g. 
- `map(string)` becomes a KeyValue component
- `list(map(string))` would become a repeater containing a KeyValue component in the schema

The forms are not tied to any model but instead the form data is encoded as json after submission.

### Problem

The repeater component requires schema components to have a Key Name, and we do not want these key names to appear in the output json.

For example, for `list(map(string))` our repeater contains a KeyValue with name "listname".
The form input, json output and ideal output are shown following:
![image](https://github.com/filamentphp/filament/assets/65358938/26e77723-759e-4856-a199-f39ac0fd2113)
```json
// Output
[{"listname":{"aa":"aa"}},{"listname":{"bb":"bb"}}]

// Target format
[{"aa":"aa"},{"bb":"bb"}]

```

We have tried several functions in `dehydrateStateUsing` to try and remove the keyname, but all of our attempts include the old unmodified data in some form, in additional to the value returned by the dehydration function.

For example, this was an attempt to extract the "listname" key if it exists:

```php
->dehydrateStateUsing(function ($state) {
    return array_map( 
        fn($item) => array_key_exists("listname", $item) ? $item["listname"] : $item,
        $state
    );
})
```
which returns the result
```json
[{"aa":"aa","listname":{"aa":"aa"}},{"bb":"bb","listname":{"bb":"bb"}}]
```

Even when we set the function to explicitly return a completely different value, it still includes the old values:

```php
->dehydrateStateUsing(function ($state) {
     return ["Whatever"];
})
```
returns the result
```json
["Whatever", {"listname":{"aa":"aa"}},{"listname":{"bb":"bb"}}]
```

My questions are:
1. Why is the dehydrateStateUsing function behaving like this for the Repeater component?
2. Is there any correct or alternate method for removing key names like this?

**Comment by danharrin** [comment, marked as answer]
The alternative could be just modifying the array returned from `getState()` before you save?

**Reply by RichardHuntCCI** [reply]
Cheers Dan,

Ended up trying that shortly after posting and it worked nicely as part of the form submission.

For anyone's future reference, our function looked like this:

```php

// This is the name given to child components of the repeater
public $repeaterName = "listname";

public function submit(): void
{
    $formState = $this->form->getState();
    $formState = $this->dehydrateRepeaterKeys($formState);
    ...
}


private function dehydrateRepeaterKeys($data) {
    if (!is_array($data)) {
        return $data;
    } elseif (count($data) == 1 && array_key_exists($this->repeaterName, $data)) {
        // Removes the $repeaterName key
        return $this->dehydrateRepeaterKeys($data[$this->repeaterName]);
    } else {
        $newData = array();
        foreach ($data as $key => $value) {
            if (is_array($value)) {
                $newData[$key] = $this->dehydrateRepeaterKeys($value);
            } else {
                $newData[$key] = $value;
            }
        }
        return $newData;
    }
}
```

We realised then that the next challenge was if we wanted to re-edit the value, then after decoding back from json, we'd have to re-insert the `listname` keys correctly for them to be passed to the correct components or otherwise cache the original untransformed value somewhere. 
That's still something we're working on so far, but beyond the scope of this question (and would be a challenge even if we used the dehydrate function).

I'd still be curious as to why the dehydrate function is working like it did in the original post, if you can offer any answers?

**Reply by danharrin** [reply]
Honestly I'd have to properly dig deep to find out why and I'm quite tight for time ATM, but there is probably a good reason

**Reply by Eighke** [reply]
I had the exact same problem, I `dehydrated(false)` my `KeyValue` and it did work. After digging everything inside the Repeater is regarded as interdependent fields and added to the state.

If you `dump($this->getStateToDehydrate())` inside `HasState::dehydrateState()` you can see the subfields being added. So in the example `data` Repeater has a `listname` Field.

```php
->dehydrateStateUsing(function ($state) {
     return ["Whatever"];
})
```

The dump will return.

```
data.data =  ["Whatever"]
data.data.0.listname = ...
data.data.1.listname = ...
```

And `Arr::set($state, $key, $value);` will just recreate a new `data.data = ["Whatever", ["listname" => ...], ["listname" => ...]]` 

---------

That say I switched `mutateDehydratedStateUsing` instead, to avoid using `dehydrated(false)`. I'm just not sure what the correct behaviour. :)
